docs/modules/otaupgrade.md

```diff
@@ -0,0 +1,176 @@
+
+The OTA Upgrade module provides access to the IDF Over-The-Air Upgrade
+support, enabling new application firmware to be applied and booted into.
+
+This module is not concerned with where the new application comes from.
+The choice of download source and method (e.g. https, tftp) is left to
+the user.
+
+In order to use the `otaupgrade` module, there must exist at least two
+OTA partitions (type `app`, subtype `ota_0` / `ota_1`), as well as the
+"otadata" partition (type `data`, subtype `ota`). The IDF implements
+the typical "flip-flop" approach to upgrades, in that one of the
+partitions hosts the running application, and the upgrade is downloaded
+into the inactive partition and only when fully downloaded and verified
+is it marked as bootable. This makes the system resilient to incomplete
+upgrades, be it due to power-loss, interrupted downloads, or other such
+things.
+
+Depending on whether the installed boot loader has been built with or
+without rollback support, the upgrade process has four or three steps.
+Without rollback support, the steps are:
+
+- `otaupgrade.commence()`
+- feed the new application image into `otaupgrade.write(data)` in chunks
+- `otaupgrade.complete(1)` to finalise and reboot into the new application
+
+If the boot loader is built with rollback support, an extra step is needed
+after the new application has booted (and been tested to be "good", by
+whatever metric(s) the user chooses):
+
+- `otaupgrade.accept()` to mark this image as valid, and allow it to be
+  booted into again.
+
+# otaupgrade.commence()
+
+Wipes the spare application partition and prepares to receive the new
+application firmware.
+
+#### Syntax
+`otaupgrade.commence()`
+
+#### Parameters
+None.
+
+#### Returns
+`nil`
+
+A Lua error may be raised if the OTA upgrade cannot be commenced for some
+reason (such as due to incorrect partition setup).
+
+
+# otaupgrade.write(data)
+
+Write a chunk of application firmware data to the correct partition and
+location. Data must be streamed sequentially, the IDF does not support
+out-of-order data as would be the case from e.g. bittorrent.
+
+#### Syntax
+`otaupgrade.write(data)`
+
+#### Parameters
+- `data` a string of binary data
+
+#### Returns
+`nil`
+
+A Lua error may be raised if the data can not be written, e.g. due to the
+data not being a valid OTA image (the IDF performs some checks in this
+regard).
+
+
+# otaupgrade.complete(reboot)
+
+Finalises the upgrade, and optionally reboots into the new application
+firmware right away.
+
+#### Syntax
+`otaupgrade.complete(reboot)`
+
+#### Parameters
+- `reboot` 1 to reboot into the new firmware immediately, nil to keep running
+
+#### Returns
+`nil`
+
+A Lua error may be raised if the image does not pass validation, or no data
+has been written to the image at all.
+
+
+# otaupgrade.accept()
+
+When the installed boot loader is built with rollback support, a new
```

Comment:
Is there a way to test if the bootloader supports rollback and if we're booting a non-accepted image?